### ISSUE TITLE
Eager CUDAGraph + stream performance

### ISSUE THREAD

**Issue by BoyuanFeng** [pytorch/pytorch#161104]
In the following repro, with warmup in `block1` and `block2`, the latency of adding two cuda tensors `x+y` is 0.0013 ms. However, if we comment out either `block1` or `block2`, the latency reduces to 0.0011 ms. This is unexpected since there are `torch.cuda.synchronize()` after `block1` and `block2`.

Repro:
```python
import torch


x = torch.randn(1024, device="cuda", requires_grad=False)
y = torch.randn(1024, device="cuda", requires_grad=False)


def fn():
    return x + y


# warmup
############################# <---- Block1: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
for _ in range(100):
    fn()
############################

torch.cuda.synchronize()

with torch.cuda.stream(torch.cuda.Stream()):
    # warmup
    ############################# <---- Block2: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
    ############################# <---- comment out 1 of option1/option2 to reduce latency from 0.0013 ms to 0.0011 ms
    for _ in range(6):
        fn()
    ############################
    n_repeat = 2000

    torch.cuda.synchronize()

    # construct a cuda graph with `n_repeat` unrolled function calls to minimize
    # host overhead
    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g):
        for _ in range(n_repeat):
            fn()

    # measure time and return
    ret = []
    n_retries = 10
    for _ in range(n_retries):
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
        g.replay()
        end_event.record()
        torch.cuda.synchronize()
        measured_latency = start_event.elapsed_time(end_event) / n_repeat
        ret.append(measured_latency)

compiler_latency = sum(ret) / len(ret)
print("compiler_latency:", compiler_latency)
```

**Comment by BoyuanFeng** [comment]
This should not relate to the queue limit on gpu.

When running 100 iterations in block1 and 0 iterations in block2, the latency is 0.0011 ms.
When running 10 iterations in block1 and 10 iterations in block2, the latency is 0.0013 ms.

**Comment by ngimel** [comment]
Can't repro, getting 0.0014 ms regardless

**Comment by BoyuanFeng** [comment]
Could you try these two scripts? I repro'ed on both H100 and B200.. 

On H100, `python3 script1.py` gives 0.00146 ms.

<details>
<summary>script1.py</summary>

```python
import torch


x = torch.randn(1024, device="cuda", requires_grad=False)
y = torch.randn(1024, device="cuda", requires_grad=False)


def fn():
    return x + y


# warmup
############################# <---- Block1: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
for _ in range(100):
    fn()
############################

torch.cuda.synchronize()

with torch.cuda.stream(torch.cuda.Stream()):
    # warmup
    ############################# <---- Block2: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
    ############################# <---- comment out 1 of option1/option2 to reduce latency from 0.0013 ms to 0.0011 ms
    for _ in range(6):
        fn()
    ############################
    n_repeat = 2000

    torch.cuda.synchronize()

    # construct a cuda graph with `n_repeat` unrolled function calls to minimize
    # host overhead
    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g):
        for _ in range(n_repeat):
            fn()

    # measure time and return
    ret = []
    n_retries = 10
    for _ in range(n_retries):
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
        g.replay()
        end_event.record()
        torch.cuda.synchronize()
        measured_latency = start_event.elapsed_time(end_event) / n_repeat
        ret.append(measured_latency)

compiler_latency = sum(ret) / len(ret)
print("compiler_latency:", compiler_latency)

```

</details>

`python script2.py` gives 0.00128 ms.


<details>
<summary>script2.py</summary>

```python
import torch


x = torch.randn(1024, device="cuda", requires_grad=False)
y = torch.randn(1024, device="cuda", requires_grad=False)


def fn():
    return x + y


# warmup
############################# <---- Block1: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
for _ in range(100):
    fn()
############################

torch.cuda.synchronize()

with torch.cuda.stream(torch.cuda.Stream()):
    # warmup
    ############################# <---- Block2: comment out this line to reduce latency from 0.0013 ms to 0.0011 ms
    ############################# <---- comment out 1 of option1/option2 to reduce latency from 0.0013 ms to 0.0011 ms
    # for _ in range(6):        ################################### <- the only change
    #     fn()
    ############################
    n_repeat = 2000

    torch.cuda.synchronize()

    # construct a cuda graph with `n_repeat` unrolled function calls to minimize
    # host overhead
    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g):
        for _ in range(n_repeat):
            fn()

    # measure time and return
    ret = []
    n_retries = 10
    for _ in range(n_retries):
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
        g.replay()
        end_event.record()
        torch.cuda.synchronize()
        measured_latency = start_event.elapsed_time(end_event) / n_repeat
        ret.append(measured_latency)

compiler_latency = sum(ret) / len(ret)
print("compiler_latency:", compiler_latency)
```

</details>

**Comment by galv** [comment]
I see. @BoyuanFeng Please give me the output of nvidia-smi so I can make sure I repro with the right driver.

**Comment by BoyuanFeng** [comment]
@galv Thanks! I repro'ed on 2 servers.

Server 1:
 NVIDIA-SMI 570.124.06             Driver Version: 570.124.06     CUDA Version: 12.9  

Server 2:
 NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4

**Comment by ngimel** [comment]
Yep, I can repro with the different scripts, my original attempts were in ipython. That's super weird (driver 550)

**Comment by eqy** [comment]
@tfogal pointed out that the excessive use of `torch.cuda.synchronize` can introduce artifacts if the GPU is ramping up and down power states, so e.g.,  `nvidia-smi --lock-gpu-clocks=MIN_CLOCK,MAX_CLOCK` to lock clocks might be useful to check
